### PR TITLE
Update username

### PR DESCRIPTION
**Pull request by zjhmale**
now can update someone's username

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,11 +6,13 @@ home
 /data
 /native
 /target
+.DS_Store
```
Comment (dotemacs) [inline .gitignore]: This should go into your own ~/.gitignore

File: .gitignore [hunk 2]
```diff
@@ -6,11 +6,13 @@ home
...
 
 # IntelliJ IDEA (La Clojure) project files
 *.ipr
 *.iws
 *.iml
+.idea
```
Comment (dotemacs) [inline .gitignore]: The same as the above comment.

File: .gitignore (hunk 1 above)
Comment (zjhmale) [inline .gitignore]: ok fix it now :)

**Comment by zjhmale** [comment]
the UT log in travis is `actual: java.io.FileNotFoundException: data/test/authorized_keys (No such file or directory)` it is really wired :cold_sweat:

**Comment by zjhmale** [comment]
done!

**Comment by tobias** [comment]
@zjhmale thanks for implementing this! I'll try to take a look and give you some feedback.

**Comment by zjhmale** [comment]
ok :)

**Comment by tobias** [comment]
What I meant to say was "try to take a look **this week** and give you feedback.".

**Comment by zjhmale** [comment]
@tobias it's fine, just take your time.

**Comment by tobias** [comment]
This looks really good, with one issue: the clojars db is a bit de-normalized, so the username is used in several tables instead of the numeric id. This means that in addition to calling `update-user`, you also need to update `jars:user`, `groups:user`, and `groups:added_by`. It might be worthwhile to introduce a `update-username` fn in `clojars.db` that does all that at once.

**Comment by zjhmale** [comment]
@tobias ok, i will get this done!

**Comment by zjhmale** [comment]
poke @tobias

**Comment by tobias** [comment]
I finally took a closer look at this (and ran it locally), and I think
it needs a couple more changes if you are willing to do them:

* we need to validate the username to confirm that it is valid (not already taken, valid characters, not reserved, etc - see https://github.com/ato/clojars-web/blob/master/src/clojars/web/user.clj#L86)
* can you merge all of these commits down to one? It's fine to force-push to this branch

Thanks!

**Comment by zjhmale** [comment]
done with validate username when update and squansh commits down to one @tobias

**Comment by xeqi** [comment]
The username is also stored in the session as the currently logged in user. It needs to be updated, or cleared, when the new name is assigned. I'm not sure friend has something easily accessible to do this. It might require `update-profile` returning a response with a `:session` containing `:cemerick.friend/identity` pointing to the new name.

If it is updated, then going back to `/profile` instead of `/login` might work.

Reproduction: create a user, login, change password, and then go to /profile.

**Comment by xeqi** [comment]
Additionally any concurrent sessions for the same user would have similar problems.  However, attempting to update/clear them would require threading the session store through the app. I'd expect this to occur rarely, and leave it to others to decide if it should be worried about.

**Comment by tobias** [comment]
@xeqi - thanks for your input.

@zjhmale - I played with this a bit over the weekend, cleaning up a
few things, and adding tests, when I ran into an issue we have not
considered: how do we deal with the `org.clojars.username` group?

I see a few options:

1) rename the group to `org.clojars.newusername`. This would also
   require moving contents of the old group to the new group name in
   the repo, and the poms for those projects would no longer be
   correct. Also, any project that depended on projects in the old
   group would break. I don't think this is a good solution.

2) leave `org.clojars.oldusername`, and create
   `org.clojars.newusername`, with the user now owning both. This
   solves the issues from (1), but adds complexity to the code for
   creating/changing usernames: we now need to check for the existence
   of a `org.clojars` group for the username even when no user with
   that name exists, since there could have been one in the past. This
   means that when someone gives up a username, it won't be available
   for anyone else to use, but that's probably okay.

3) same as (2), but only leave `org.clojars.oldusername` if it
   actually has artifacts deployed under it. I don't think this is
   worth doing, since it just adds complexity to the code and
   confusion for users, since changing the username can now sometimes
   free up the old name, and the reason why may not be evident.

Of those three options, I think (2) is the best. But thinking about
the complexity it adds makes me want to take a step back and consider
the complexity vs. utility trade-off of this entire feature. I'm not
rejecting this yet, but I would like to discuss the motivations for
this change. How often to users need/want to change their usernames?
What are situations where changing the username would be needed?

I really appreciate the work you have put into this, and apologize for
not discussing motivations sooner, before asking you to make changes
to your PR. Like I said above, I'm not quite ready to reject this, and
would like input on how useful this feature really is.

For reference, I have parked the commit from this PR (along with my
additional changes) on the [update-username branch](https://github.com/ato/clojars-web/tree/update-username).

**Comment by zjhmale** [comment]
hey @tobias  this is really a big issue i did not expected either, there is a default group for every user named `org.clojars.username`. and for the second plan, there definately a lot work to do.

for myself, i just want to update a cooler name on clojars, i did not expect the username is related to a specific group name.

maybe i can implement plan B this week and after that, we can review it to decide whether add this feature to main branch.

**Comment by tobias** [comment]
I'm closing this since it has turned out have a bigger impact that we originally thought. Thanks again for all of your effort!

**Pull request closed without merging**